flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/dml/RichSqlInsert.java

```diff
@@ -0,0 +1,292 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.sql.parser.dml;
+
+import org.apache.flink.sql.parser.SqlProperty;
+import org.apache.flink.sql.parser.ddl.ExtendedSqlNode;
+import org.apache.flink.sql.parser.error.SqlParseException;
+
+import org.apache.calcite.plan.RelOptTable;
+import org.apache.calcite.prepare.Prepare;
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.rel.type.RelDataTypeFactory;
+import org.apache.calcite.rel.type.RelDataTypeField;
+import org.apache.calcite.sql.SqlIdentifier;
+import org.apache.calcite.sql.SqlInsert;
+import org.apache.calcite.sql.SqlLiteral;
+import org.apache.calcite.sql.SqlNode;
+import org.apache.calcite.sql.SqlNodeList;
+import org.apache.calcite.sql.SqlSelect;
+import org.apache.calcite.sql.SqlUtil;
+import org.apache.calcite.sql.SqlWriter;
+import org.apache.calcite.sql.fun.SqlStdOperatorTable;
+import org.apache.calcite.sql.parser.SqlParserPos;
+import org.apache.calcite.sql.type.SqlTypeUtil;
+import org.apache.calcite.sql.validate.SqlValidator;
+import org.apache.calcite.sql.validate.SqlValidatorNamespace;
+import org.apache.calcite.sql.validate.SqlValidatorScope;
+import org.apache.calcite.sql.validate.SqlValidatorTable;
+import org.apache.calcite.sql.validate.SqlValidatorUtil;
+import org.apache.calcite.util.Pair;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import static org.apache.calcite.util.Static.RESOURCE;
+
+/** A {@link SqlInsert} that have some extension functions like partition, overwrite. **/
+public class RichSqlInsert extends SqlInsert implements ExtendedSqlNode {
+	private final SqlNodeList staticPartitions;
+
+	public RichSqlInsert(SqlParserPos pos,
+			SqlNodeList keywords,
+			SqlNode targetTable,
+			SqlNode source,
+			SqlNodeList columnList,
+			SqlNodeList staticPartitions) {
+		super(pos, keywords, targetTable, source, columnList);
+		this.staticPartitions = staticPartitions;
+	}
+
+	@Override
+	public void validate(SqlValidator validator, SqlValidatorScope scope) {
+		final SqlNode source = this.getSource();
+		if (this.staticPartitions != null
+			&& this.staticPartitions.size() > 0
+			&& source instanceof SqlSelect) {
+			appendPartitionProjects(validator, (SqlSelect) source, this.staticPartitions);
```

Comment:
Looks like a hack. Normally `validate` should be idempotent, do we have other choices to achieve what you want to do?

Reply:
I don't think it is a hack, we need a full schema(plus the static partitions) of source table for validation, while some engines like Spark or Hive only allow the partition columns to be the last columns of the target table, we do not have this limitation.

Reply:
What if we `validate` twice?

Reply:
We only validate once for SqlValidator, maybe plus our custom validation defined in `ExtendedSqlNode`

Reply:
I think we shouldn't rely on this unreliable contract. For a well defined interface, you need to be self contained. Mostly, interfaces named like `validate` will not have side effects, and can be called multiple times.

Reply:
I agree, have added self check to ensure that this side-effects only take once.